File: src/bcd_web_vue/js/models/item.js
/**
 * Item and bibliographic record data models with TypeScript JSDoc type definitions.
 * Fully aligned with Pydantic and SQLAlchemy models.
 */

/**
 * Basic bibliographic record information
 * @typedef {Object} BibliographicRecord
 * @property {number} id - Bibliographic record database ID
 * @property {string|null} isbn - ISBN-13 or ISBN-10
 * @property {string} title - Book title
 * @property {string[]|null} authors - List of authors
 * @property {string|null} publisher - Publisher name
 * @property {number|null} publication_year - Year of publication
 * @property {string|null} language - Language code (e.g., 'fr', 'en')
 * @property {string|null} target_audience - Target audience
 * @property {string|null} summary - Book summary/description
 * @property {string[]|null} subjects - Subject keywords
 * @property {string|null} medium_type - Medium/support type
 * @property {string|null} collection - Series or collection name
 * @property {string|null} dewey_number - Dewey decimal number
 * @property {number} total_copies - Total number of physical copies
 * @property {number} available_copies - Number of available copies
 */

/**
 * Physical item/copy information
 * @typedef {Object} Item
 * @property {number} id - Item database auto-increment ID
 * @property {string} item_id - Unique alphanumeric item ID
 * @property {number} bibliographic_record_id - Parent bibliographic record ID
 * @property {string} barcode - Item barcode (equal to item_id)
 * @property {string|null} call_number - Call number (Dewey/CDU)
 * @property {string|null} shelf_location - Physical shelf location
 * @property {string} condition - Condition: 'good', 'damaged'
 * @property {string} status - Status: 'available', 'on_loan', 'on_hold', 'in_repair', 'lost', 'withdrawn'
 * @property {boolean} loanable - Whether the item can be borrowed
 * @property {string|null} acquisition_date - Date acquired (ISO 8601)
 * @property {string|null} funding_source - Funding source
 * @property {string|null} last_borrowed_at - Timestamp of last checkout (ISO 8601)
 * @property {string|null} last_inventoried_at - Timestamp of last inventory scan (ISO 8601)
 */

/**
 * Detailed item with current loan information
 * @typedef {Object} ItemDetailed
 * @property {number} id - Item database ID
 * @property {string} item_id - Item identifier
 * @property {string|null} call_number - Call number
 * @property {string|null} shelf_location - Physical shelf location
 * @property {string} status - Item status
 * @property {string} condition - Item condition
 * @property {boolean} loanable - Whether item is loanable
 * @property {string|null} acquisition_date - Date acquired (ISO 8601)
 * @property {string|null} funding_source - Funding source
 * @property {Object|null} current_loan - Current loan info (null if available)
 * @property {string} current_loan.borrower_id - Borrower ID
 * @property {string} current_loan.borrower_name - Borrower name
 * @property {string} current_loan.due_date - Due date (ISO 8601)
 * @property {boolean} current_loan.is_overdue - Whether overdue
 * @property {number} current_loan.days_overdue - Days overdue
 */

/**
 * Bibliographic record with all items/copies
 * @typedef {Object} BibliographicRecordWithItems
 * @property {number} id - Bibliographic record ID
 * @property {string|null} isbn - ISBN
 * @property {string} title - Book title
 * @property {string[]|null} authors - List of authors
 * @property {string|null} publisher - Publisher name
 * @property {number|null} publication_year - Year of publication
 * @property {string|null} language - Language code
 * @property {string|null} target_audience - Target audience
 * @property {string|null} summary - Book summary
 * @property {string[]|null} subjects - Subject keywords
 * @property {string|null} medium_type - Medium type
 * @property {string|null} collection - Series/collection
 * @property {string|null} dewey_number - Dewey decimal number
 * @property {number} total_copies - Total number of copies
 * @property {number} available_copies - Available copies
 * @property {ItemDetailed[]} items - Array of all physical items/copies with detailed loan info
 */

// Export empty object to make this a module
export {};

File: src/bcd_api/schemas/item.py
"""Pydantic schemas for Item model."""

from typing import Optional
from datetime import date, datetime
from pydantic import BaseModel, ConfigDict, Field

from src.shared.constants import ItemStatus, ItemCondition
from src.bcd_api.schemas.common import TimestampMixin


class ItemBase(BaseModel):
    """Base schema for Item."""

    item_id: str = Field(..., min_length=1, max_length=20, description="Unique item ID")
    bibliographic_record_id: int = Field(..., description="Bibliographic record ID")
    call_number: Optional[str] = Field(None, max_length=50, description="Call number (Dewey/CDU)")
    shelf_location: Optional[str] = Field(None, max_length=100, description="Physical shelf location")
    condition: ItemCondition = Field(ItemCondition.GOOD, description="Physical condition")
    status: ItemStatus = Field(ItemStatus.AVAILABLE, description="Availability status")
    loanable: bool = Field(True, description="Can be borrowed?")
    acquisition_date: Optional[date] = Field(None, description="Acquisition date")
    funding_source: Optional[str] = Field(None, max_length=100, description="Funding source")


class ItemCreate(ItemBase):
    """Schema for creating a new item."""

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "item_id": "785",
                "bibliographic_record_id": 1,
                "call_number": "800.000",
                "shelf_location": "Fiction - Section A - Row 3",
                "loanable": True,
                "acquisition_date": "2024-09-15",
                "funding_source": "Budget 2024-2025",
            }
        }
    )


class ItemUpdate(BaseModel):
    """Schema for updating an item."""

    call_number: Optional[str] = Field(None, max_length=50)
    shelf_location: Optional[str] = Field(None, max_length=100)
    condition: Optional[ItemCondition] = None
    status: Optional[ItemStatus] = None
    loanable: Optional[bool] = None
    acquisition_date: Optional[date] = None
    funding_source: Optional[str] = Field(None, max_length=100)


class ItemResponse(ItemBase, TimestampMixin):
    """Schema for item response."""

    id: int
    barcode: str
    last_borrowed_at: Optional[datetime]

    model_config = ConfigDict(from_attributes=True)


class ItemSummary(BaseModel):
    """Summary schema for item (for lists)."""

    id: int
    item_id: str
    barcode: str
    call_number: Optional[str]
    status: ItemStatus
    condition: ItemCondition
    loanable: bool

    model_config = ConfigDict(from_attributes=True)


class ItemWithBiblio(ItemResponse):
    """Item with bibliographic record information."""

    title: str = Field(..., description="Title from bibliographic record")
    authors: Optional[list[str]] = Field(None, description="Authors")

    model_config = ConfigDict(from_attributes=True)


class CurrentLoanInfo(BaseModel):
    """Current loan information for an item."""

    borrower_id: str
    borrower_name: str
    due_date: str  # Formatted date string
    is_overdue: bool
    days_overdue: int


class ItemWithCurrentLoan(BaseModel):
    """Item with current loan information (for catalog display)."""

    id: int  # Database ID needed for editing
    item_id: str
    call_number: Optional[str]
    shelf_location: Optional[str]
    status: ItemStatus
    condition: ItemCondition
    loanable: bool
    acquisition_date: Optional[date] = None
    funding_source: Optional[str] = None
    current_loan: Optional[CurrentLoanInfo] = None

    model_config = ConfigDict(from_attributes=True)


class AvailableIDsResponse(BaseModel):
    """Response schema for available item IDs endpoint."""

    start_id: str = Field(..., description="First ID in the generated range")
    end_id: str = Field(..., description="Last ID in the generated range")
    ids: list[str] = Field(..., description="List of available item IDs")
    count: int = Field(..., description="Number of IDs generated")
    id_format: str = Field(..., description="ID format (numeric or alphanumeric)")
    contiguous: bool = Field(True, description="Whether IDs form a gapless block")

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "start_id": "2000",
                "end_id": "2029",
                "ids": ["2000", "2001", "2002"],
                "count": 30,
                "id_format": "numeric",
                "contiguous": True
            }
        }
    )
